### PR TITLE
changed: move copying of refinement to separate method

### PR DESCRIPTION
**Pull request by akva2**
to facilitate reuse elsewhere

### CODE REVIEW

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Comment by akva2** [comment]
jenkins build this with downstreams please

File: src/ASM/LR/ASMu2Dmx.C [hunk 1]
```diff
@@ -993,28 +993,12 @@ bool ASMu2Dmx::refine (const LR::RefineData& prm, Vectors& sol)
...
-                                            line->start_, line->stop_, mult);
-          else
-            m_basis[j]->insert_const_u_edge(line->const_par_,
-                                            line->start_, line->stop_, mult);
-        }
+    for (size_t j = 0; j < m_basis.size(); ++j)
+      if (refBasis != m_basis[j]) {
+        int mult = (j == 0 && ASMmxBase::Type == REDUCED_CONT_RAISE_BASIS1) ||
+                   (j == 1 && ASMmxBase::Type == REDUCED_CONT_RAISE_BASIS2) ? 2 : 1;
+        this->copyRefinement(m_basis[j].get(), mult);
```
Comment (kmokstad) [inline src/ASM/LR/ASMu2Dmx.C:1000]: Cosmetic, but maybe the following would be as readable?
```
  if ((j == 0 && ASMmxBase::Type == REDUCED_CONT_RAISE_BASIS1) ||
      (j == 1 && ASMmxBase::Type == REDUCED_CONT_RAISE_BASIS2))
    this->copyRefinement(m_basis[j].get(),2);
  else
    this->copyRefinement(m_basis[j].get(),1);

File: src/ASM/LR/ASMu2Dmx.C [hunk 2]
```diff
@@ -1192,3 +1176,23 @@ void ASMu2Dmx::storeMesh (const std::string& fName, int fType) const
   writeBasis(projBasis.get(), "proj");
   writeBasis(refBasis.get(), "ref");
 }
+
+
+void ASMu2Dmx::copyRefinement(LR::LRSplineSurface* basis,
+                              int multiplicity) const
+{
+  for (const LR::Meshline* line : refBasis->getAllMeshlines()) {
+    int mult = 1;
```
Comment (kmokstad) [inline src/ASM/LR/ASMu2Dmx.C:1185]: Suggestion:
```
int mult = line->multiplicity_ > 1? line->multiplicity_  : multiplicity;
```

File: src/ASM/LR/ASMu3Dmx.C [hunk 3]
```diff
@@ -924,26 +924,12 @@ bool ASMu3Dmx::refine (const LR::RefineData& prm, Vectors& sol)
...
-          LR::MeshRectangle* newRect = rect->copy();
-          newRect->multiplicity_ = mult;
-
-          m_basis[j]->insert_line(newRect);
-        }
+    for (size_t j = 0; j < m_basis.size(); ++j)
+      if (refBasis != m_basis[j]) {
+        int mult = (j == 0 && ASMmxBase::Type == REDUCED_CONT_RAISE_BASIS1) ||
+                   (j == 1 && ASMmxBase::Type == REDUCED_CONT_RAISE_BASIS2) ? 2 : 1;
+        this->copyRefinement(m_basis[j].get(), mult);
```
Comment (kmokstad) [inline src/ASM/LR/ASMu3Dmx.C:931]: Same here?

File: src/ASM/LR/ASMu3Dmx.C [hunk 4]
```diff
@@ -1039,3 +1025,21 @@ size_t ASMu3Dmx::getNoRefineElms() const
...
+
+void ASMu3Dmx::copyRefinement(LR::LRSplineVolume* basis, int multiplicity) const
+{
+  for (const LR::MeshRectangle* rect : refBasis->getAllMeshRectangles()) {
+    int p = basis->order(rect->constDirection());
+    int mult = 1;
+    if (rect->multiplicity_ > 1)
+      mult = p;
+    else
+      mult = multiplicity;
```
Comment (kmokstad) [inline src/ASM/LR/ASMu3Dmx.C:1038]: Similar here (2nd comment)

File: src/ASM/LR/ASMu2Dmx.C (hunk 1 above)
Comment (akva2) [reply src/ASM/LR/ASMu2Dmx.C]: done.

File: src/ASM/LR/ASMu3Dmx.C (hunk 3 above)
Comment (akva2) [reply src/ASM/LR/ASMu3Dmx.C]: done

File: src/ASM/LR/ASMu3Dmx.C (hunk 4 above)
Comment (akva2) [reply src/ASM/LR/ASMu3Dmx.C]: done

File: src/ASM/LR/ASMu2Dmx.C (hunk 2 above)
Comment (akva2) [reply src/ASM/LR/ASMu2Dmx.C]: done

**Comment by akva2** [comment]
jenkins build this with downstreams please

**Pull request merged**